File: packages/expo-file-system/ios/Next/FileSystemPath.swift
import Foundation
import ExpoModulesCore

internal class FileSystemPath: SharedObject {
  var url: URL

  init(url: URL, isDirectory: Bool) {
    let standardizedUrl = url.deletingLastPathComponent().appendingPathComponent(url.lastPathComponent, isDirectory: isDirectory)
    self.url = standardizedUrl
  }

  func validatePermission(_ flag: EXFileSystemPermissionFlags) throws {
    try ensurePathPermission(appContext, path: url.path, flag: flag)
  }

  func validateCanCreate(_ options: CreateOptions) throws {
    if try !options.overwrite && exists {
      throw FileAlreadyExistsException("File already exists")
    }
  }

  var exists: Bool {
    get throws {
      FileManager.default.fileExists(atPath: url.path)
    }
  }

  func delete() throws {
    try validatePermission(.write)
    guard FileManager.default.fileExists(atPath: url.path) else {
      throw UnableToDeleteException("path does not exist")
    }
    do {
      try FileManager.default.removeItem(at: url)
    } catch {
      throw UnableToDeleteException(error.localizedDescription)
    }
  }

  func getMoveOrCopyPath(to destination: FileSystemPath) throws -> URL {
    if let destination = destination as? FileSystemDirectory {
      if self is FileSystemFile {
        return destination.url.appendingPathComponent(url.lastPathComponent)
      }
      // self if FileSystemDirectory
      // we match unix behavior https://askubuntu.com/a/763915
      if destination.exists {
        return destination.url.appendingPathComponent(url.lastPathComponent, isDirectory: true)
      }
      return destination.url
    }
    // destination is FileSystemFile
    guard self is FileSystemFile else {
      throw CopyOrMoveDirectoryToFileException()
    }
    return destination.url
  }

  func copy(to destination: FileSystemPath) throws {
    try validatePermission(.read)
    try destination.validatePermission(.write)
    try FileManager.default.copyItem(at: url, to: getMoveOrCopyPath(to: destination))
  }

  func move(to destination: FileSystemPath) throws {
    try validatePermission(.write)
    try destination.validatePermission(.write)
    let destinationUrl = try getMoveOrCopyPath(to: destination)
    try FileManager.default.moveItem(at: url, to: destinationUrl)
    url = destinationUrl
  }
}

File: packages/expo-file-system/android/src/main/java/expo/modules/filesystem/next/FileSystemPath.kt
package expo.modules.filesystem.next

import android.os.Build
import expo.modules.interfaces.filesystem.Permission
import expo.modules.kotlin.sharedobjects.SharedObject
import java.io.File
import java.net.URI
import java.nio.file.attribute.BasicFileAttributes
import java.util.EnumSet
import kotlin.io.path.Path
import kotlin.io.path.moveTo
import kotlin.io.path.readAttributes
import kotlin.time.Duration.Companion.milliseconds

abstract class FileSystemPath(public var uri: URI) : SharedObject() {
  val file get() = File(uri)

  fun delete(fileOrDirectory: File = file) {
    if (!fileOrDirectory.exists()) {
      throw UnableToDeleteException("path '${fileOrDirectory.path}' does not exist")
    }
    if (fileOrDirectory.isDirectory) {
      fileOrDirectory.listFiles()?.forEach { child ->
        if (child.isDirectory) {
          // Recursively delete subdirectories
          delete(child)
        } else {
          if (!child.delete()) {
            throw UnableToDeleteException("failed to delete '${child.path}'")
          }
        }
      }
    }
    if (!fileOrDirectory.delete()) {
      throw UnableToDeleteException("failed to delete '${fileOrDirectory.path}'")
    }
  }

  abstract fun validateType()

  fun getMoveOrCopyPath(destination: FileSystemPath): File {
    if (destination is FileSystemDirectory) {
      if (this is FileSystemFile) {
        if (!destination.exists) {
          throw DestinationDoesNotExistException()
        }
        return File(destination.file, file.name)
      }
      // this if FileSystemDirectory
      // we match unix behavior https://askubuntu.com/a/763915
      if (destination.exists) {
        return File(destination.file, file.name)
      }
      if (destination.file.parentFile?.exists() != true) {
        throw DestinationDoesNotExistException()
      }
      return destination.file
    }
    // destination is FileSystemFile
    if (this !is FileSystemFile) {
      throw CopyOrMoveDirectoryToFileException()
    }
    if (destination.file.parentFile?.exists() != true) {
      throw DestinationDoesNotExistException()
    }
    return destination.file
  }

  fun validatePermission(permission: Permission) {
    if (!checkPermission(permission)) {
      throw InvalidPermissionException(permission)
    }
  }

  fun checkPermission(permission: Permission): Boolean {
    val permissions = appContext?.filePermission?.getPathPermissions(appContext?.reactContext, file.path) ?: EnumSet.noneOf(Permission::class.java)
    return permissions.contains(permission)
  }

  fun validateCanCreate(options: CreateOptions) {
    if (!options.overwrite && file.exists()) {
      throw UnableToCreateException("it already exists")
    }
  }

  fun copy(to: FileSystemPath) {
    validateType()
    to.validateType()
    validatePermission(Permission.READ)
    to.validatePermission(Permission.WRITE)

    file.copyRecursively(getMoveOrCopyPath(to))
  }

  fun move(to: FileSystemPath) {
    validateType()
    to.validateType()
    validatePermission(Permission.WRITE)
    to.validatePermission(Permission.WRITE)

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val destination = getMoveOrCopyPath(to)
      file.toPath().moveTo(destination.toPath())
      uri = destination.toURI()
    } else {
      file.copyTo(getMoveOrCopyPath(to))
      file.delete()
      uri = getMoveOrCopyPath(to).toURI()
    }
  }

  val modificationTime: Long get() {
    validateType()
    return file.lastModified()
  }

  val creationTime: Long? get() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      validateType()
      val attributes = Path(file.path).readAttributes<BasicFileAttributes>()
      return attributes.creationTime().toMillis().milliseconds.inWholeMilliseconds
    } else {
      return null
    }
  }
}
